packages/@glimmer/reference/lib/reference.ts

```diff
@@ -9,3 +10,128 @@ export default Reference;
 export interface PathReference<T> extends Reference<T> {
   get(key: string): PathReference<Opaque>;
 }
+
+//////////
+
+export interface VersionedReference<T = Opaque> extends Reference<T>, Tagged {}
+
+export interface VersionedPathReference<T = Opaque> extends PathReference<T>, Tagged {
+  get(property: string): VersionedPathReference<Opaque>;
+}
+
+export abstract class CachedReference<T> implements VersionedReference<T> {
+  public abstract tag: Tag;
+
+  private lastRevision: Option<Revision> = null;
+  private lastValue: Option<T> = null;
+
+  value(): T {
+    let { tag, lastRevision, lastValue } = this;
+
+    if (lastRevision === null || !validate(tag, lastRevision)) {
+      lastValue = this.lastValue = this.compute();
+      this.lastRevision = value(tag);
+    }
+
+    return lastValue as T;
+  }
+
+  protected abstract compute(): T;
+
+  protected invalidate() {
+    this.lastRevision = null;
+  }
+}
+
+//////////
+
+export type Mapper<T, U> = (value: T) => U;
+
+class MapperReference<T, U> extends CachedReference<U> {
+  public tag: Tag;
+
+  private reference: VersionedReference<T>;
+  private mapper: Mapper<T, U>;
+
+  constructor(reference: VersionedReference<T>, mapper: Mapper<T, U>) {
+    super();
+    this.tag = reference.tag;
+    this.reference = reference;
+    this.mapper = mapper;
+  }
+
+  protected compute(): U {
+    let { reference, mapper } = this;
+    return mapper(reference.value());
+  }
+}
+
+export function map<T, U>(
+  reference: VersionedReference<T>,
+  mapper: Mapper<T, U>
+): VersionedReference<U> {
+  return new MapperReference<T, U>(reference, mapper);
+}
+
+//////////
+
+export class ReferenceCache<T> implements Tagged {
+  public tag: Tag;
+
+  private reference: VersionedReference<T>;
+  private lastValue: Option<T> = null;
+  private lastRevision: Option<Revision> = null;
+  private initialized = false;
+
+  constructor(reference: VersionedReference<T>) {
+    this.tag = reference.tag;
+    this.reference = reference;
+  }
+
+  peek(): T {
```

Comment:
why does peek() force it compute? that smells a bit.

Reply:
@pzuraq What's up here?

Reply:
This code was also ported over without changes from the `validators.ts` file. I agree it probably doesn't make sense to force a compute from a method called `peek`, but I do think that's outside of the scope of this PR/change, happy to address it in a followup though when we start cleaning some stuff up